Add ICD support

jrprice opened the pull request:
All OpenCL API objects now inherit from an `icd_object` base class which contains the dispatch table. Downcasts have been introduced as necessary to convert to the clvk subclasses.

Linking to clvk directly is still supported. The dispatch table function pointers are obtained by calling `dlsym()` on the clvk library, which circumvents having to rename the API definitions to avoid clashing with the definitions provided by the ICD loader.

This is perhaps a slightly strange approach, so let me know if you'd rather I just rename the APIs and set the function pointers directly (e.g. `m_dispatch->clFoo() = clvk::clFoo()`, which then requires an another definition of `clFoo()` which calls `clvk::clFoo()` to make direct linking work). 

The many `static_cast` calls are a little ugly; I couldn't see an easy way to make them nicer. We could shorten them a bit by adding a conversion method to `icd_object`, which would lead to (e.g.) `context.as<cvk_context*>()`, but this only saves a few characters. Let me know what you think.

Tested this with OpenCL CTS -  no regressions.

kpet commented on src/kernel.hpp:
```diff
@@ -192,10 +193,13 @@ struct cvk_kernel_argument_values {
-            m_kernel_resources[arg.binding].reset(refc);
+            if (arg.kind == kernel_argument_kind::sampler) {
+                m_kernel_resources[arg.binding].reset(static_cast<cvk_sampler*>(
+                    *reinterpret_cast<const cl_sampler*>(value)));
```
I suggest doing this in two/three steps (value to api sampler, api sampler to sampler, use) for readability.

jrprice commented:
Thanks for the review. I've pushed updates which I believe address your comments.

- `-Bsymbolic` - I thought there would probably be a flag to do this, just couldn't find it! I've switched to this approach with a statically defined dispatch table, using a default initializer in the `icd_object` class so that no other code has to worry about doing this. I've also tested on macOS where it seems to work without any additional linker flags.

- I've added the `icd_downcast` routine you suggested. Note that there are still several cases where a `static_cast` is required (e.g. for downcasting `cl_mem` to buffer/image classes depending on the use).

- I've added a commit that renames function parameters. It didn't make the patch /that/ much smaller, partly due to the fact that renaming the function parameters changes the way the code is formatted. The original reason I avoided renaming the function parameters was to keep them consistent with the names used in the spec/headers. I'll let you decide whether to keep this commit or not :-)

- Added CI coverage for the ICD interface (on Linux only). I /think/ it's correct (tested on a Debian-derived workstation), but it's hard to tell if it's actually running with the ICD interface in CI.

kpet commented on src/api.cpp:
```diff
@@ -89,26 +91,7 @@ struct api_query_string : public std::string {
-    }
-
-    return CL_SUCCESS;
+    return clIcdGetPlatformIDsKHR(num_entries, platforms, num_platforms);
```
This will result in incorrect logging. I suggest introducing a `cvk_get_platform_ids` function called used by both `clGetPlatformIDs` and `clIcdGetPlatformIDsKHR`. The logging will have to be duplicated.

kpet commented on src/device.hpp:
```diff
@@ -210,9 +209,12 @@ typedef struct _cl_device_id {
+}
 
-typedef struct _cl_platform_id {
+typedef struct _cl_platform_id : public clvk::icd_object {
```
Any specific reason `cvk_platform` is declared differently from all other objects?

kpet commented on src/memory.hpp:
```diff
@@ -298,7 +306,7 @@ struct cvk_image : public cvk_mem {
     size_t depth() const { return m_desc.image_depth; }
     size_t array_size() const { return m_desc.image_array_size; }
-    cvk_mem* buffer() const { return m_desc.buffer; }
+    cvk_mem* buffer() const { return static_cast<cvk_mem*>(m_desc.buffer); }
```
Shouldn't this one be an `icd_downcast`?

kpet commented on tests/travis/step-script.sh:
```diff
@@ -46,3 +46,10 @@ fi
+
+if [ "${TRAVIS_OS_NAME}" == "linux" ]; then
+    echo $PWD/libOpenCL.so >clvk.icd
+    export OCL_ICD_VENDORS=$PWD
```
We could probably use `OCL_ICD_FILENAMES` to avoid the file. It doesn't matter really.

jrprice replied:
This environment variable is not provided by `ocl-icd`, which I believe is the ICD loader packaged by Debian (and thus what Travis CI is using).

kpet replied:
Right. We can test with the Khronos loader later (I did use that in my local testing and it worked fine so I'm not overly worried). Most distributions use `ocl-icd` at the moment anyway.

jrprice commented:
Addressed comments, PTAL.

kpet commented:
Seems you went back on the CMake in the last change. The macOS bots are failing. I think it would be more practical to avoid force-pushing in the future, I would likely have spotted that in review (you can't use the diff since last review feature when force-pushing).

jrprice commented:
Fixed now, sorry about that.

I force pushed because I had rebased on top of the latest changes on master, but that wasn't really necessary for this PR. Will avoid doing this in the future.
